### ISSUE TITLE
True TRAP_TAKEN in the EX stage causes record to be skipped without retry after TRAP

### ISSUE THREAD

**Issue by AlexxMarkov** [rafaelcalcada/steel-core#5]
Change [29235041](https://github.com/rafaelcalcada/steel-core/commit/29235041d6280fc4787ff3e8117d890b91d1b3b7) purportedly causes a CSR or REG entry to be skipped during TRAP occurrence, but does not retry the instruction after TRAP processing.

If we assume that the EX stage receives an instruction that should lead to writing to a register file or CSR, and TRAP_TAKEN occurs at the decoding stage.

Here's what will happen.

1. On the next rising edge of clk, the state machine will transition from the STATE_OPERATING stage to the STATE_TRAP_TAKEN stage.
Where SET_EPC will be set to true:

   ```
   SET_EPC = 1'b1;
   ```

   1.a. In the register PC_reg (Stages 2/3 interface registers), the value of pc from the decoding stage will be written.
The value in PC_reg will correspond to the address of the instruction on which the TRAP occurred.

   1.b. Writing to a register file or CSR will not be executed.

2. On the next rising edge of clk, the state machine will transition from the STATE_TRAP_TAKEN stage to the STATE_OPERATING stage.
   The MEPC register will be written to the PC_reg value to be returned from TRAP.

This means that after processing the TRAP vector in the execution of the mret instruction, instruction fetching is performed using the PC_reg value from the decode stage.
And an attempt to write to a register file or CSR in the EX stage (was blocked) will not be resumed.

Perhaps I made a mistake somewhere, I ask for your attention. 
I'm sorry for the inconvenience.

**Comment by rafaelcalcada** [comment]
Hello, thanks for your contribution. Indeed change [29235041](https://github.com/rafaelcalcada/steel-core/commit/29235041d6280fc4787ff3e8117d890b91d1b3b7) is making Steel return to the wrong instruction. I'll post a fix to that soon. 

The right value to be written to MEPC register is the address of the instruction in EX stage at the time the trap was taken, and not the address of the instruction in decoding stage as it is doing right now. Thanks for bringing that to attention!

**Comment by AlexxMarkov** [comment]
Is it possible to say that in the EX stage, everything should be done regardless of what happened in the decoding stage? Regardless of the occurrence of TRAP (at the stage of decoding).

At first glance, it seems that if there is an intermediate step between writing in the EX stage and executing the first instruction in TRAP and / or changing the CSR, then the behavior is defined.
The step that causes the recording in the EX stage, on the result of which the behavior of hart depends, will determine this behavior.

Steel probably has such a step, this is the STATE_TRAP_TAKEN state machine stage, passing through this state, all records at the EX stage will be completed by the time the result is used.

Except that a TRAP handler will be executed to handle the interrupt when it is enabled and when the EX stage (next rising edge of clk) resets the MIE bit or others.

Perhaps this is not so important, since after processing TRAP, Steel will continue execution with the next instruction after the one that supposedly (as an example) could reset the MIE bit.

But I may be wrong.

**Comment by rafaelcalcada** [comment]
Hi Alexx,

For an instruction that clears MIE:
- if it is in DEC stage and an interrupt/exception takes place, its execution is halted. The execution of the instruction that precedes this instruction, which is in EX stage, is also halted. The address of the execution in EX stage is saved in MEPC. When returning from the trap, the instruction in EX stage goes through DEC stage again. This is needed because some instructions, like branches and jumps, don't do any work in EX stage. Note that our instruction that clears MIE will come right after the instruction that was in EX stage, regardless of what instruction was in EX stage, so it is guaranteed that it will eventually be executed.
- if it is in EX stage and an interrupt/exception takes place, its execution is halted too. MEPC will be set to the address of this instruction, which will resume after returning from the trap.

When an interrupt/exception takes place, the instructions in DEC and EX stages have their processing halted immediately, that is, when the hart is still in OPERATING_STATE. In the next clock cycle the hart will join STATE_TRAP_TAKEN, which don't process any instructions, but make changes in hart's internal state by setting PC and some other CSR registers.

**Comment by rafaelcalcada** [comment]
Hi again,

When running the RISC-V Compliance tests I found out that the changes I made caused some tests to fail. After reading a lot about trap handling in RISC-V, I found out that there are no problems with the previous behavior. The expected behavior is exactly what it was before, so I reverted the changes.

So correcting what I said before, the expected behavior for that scenario when an instruction that clears MIE is in the pipeline is as follows:

* if the instruction that clears MIE is in DEC stage and an interrupt/exception takes place, its execution is halted. The instruction that precedes the instruction that clears MIE (which is in EX stage) is **committed**. The address of the instruction in **DEC** stage is saved in MEPC. That is, the address of the instruction that clears MIE is saved in MEPC so it can resume execution after trap handler routine finishes its execution.
* if the instruction that clears MIE is in EX stage and an interrupt/exception takes place, its results are committed. Note that in this case the interrupt is taken even if it will be disabled in the next clock cycle. What counts is the state of the core at the time the interrupt took place and not their future state. This is the correct behavior and that's why RISC-V compliance tests were failing. In the next clock cycle the value of MIE (0, because the instruction resets MIE) will be saved in MPIE and the core will execute the trap handler. MIE will restore its value (to zero) when the trap handler finish its execution (by executing mret instruction).

Sorry for the confusion! I'll close this issue soon an leave Steel as it was before. Note that issue #4 fixed an important issue related to trap handling so it is important to keep the latest version of the repo on your machine.
